app/scripts/metamask-controller.js

```diff
@@ -3494,41 +3494,29 @@ export default class MetamaskController extends EventEmitter {
         this._convertMnemonicToWordlistIndices(seedPhraseAsBuffer),
       );
 
-      // Scan accounts until we find an empty one
-      const { chainId } = this.networkController.state.providerConfig;
+      // Scan accounts and apply checks
       const ethQuery = new EthQuery(this.provider);
       const accounts = await this.keyringController.getAccounts();
-      let address = accounts[accounts.length - 1];
-
-      for (let count = accounts.length; ; count++) {
-        const balance = await this.getBalance(address, ethQuery);
+      let emptyAccountsCounter = 0;
 
-        if (balance === '0x0') {
-          // This account has no balance, so check for tokens
-          await this.detectTokensController.detectNewTokens({
-            selectedAddress: address,
-          });
-
-          const tokens =
-            this.tokensController.state.allTokens?.[chainId]?.[address];
-          const detectedTokens =
-            this.tokensController.state.allDetectedTokens?.[chainId]?.[address];
+      for (
+        let index = accounts.length - 1;
+        emptyAccountsCounter < 10;
+        index++
+      ) {
+        let address = accounts[index];
+        const isAccountEmpty = await this.isAccountEmpty(address, ethQuery);
 
-          if (
-            (tokens?.length ?? 0) === 0 &&
-            (detectedTokens?.length ?? 0) === 0
-          ) {
-            // This account has no balance or tokens
-            if (count !== 1) {
-              await this.removeAccount(address);
-            }
-            break;
+        if (isAccountEmpty) {
+          emptyAccountsCounter += 1;
+          if (index !== 0) {
+            await this.removeAccount(address);
           }
+        } else {
+          emptyAccountsCounter = 0;
+          ({ addedAccountAddress: address } =
+            await this.keyringController.addNewAccount(index + 1));
```

Comment:
I think `addNewAccount` needs to be called on every iteration, not just conditionally when the account is not empty.  Otherwise you can't proceed to check the next account since it's not added to the keyring.

I'm not super familiar with this account code, but this seemed to work for me:

```ts
      let address = accounts[accounts.length - 1];
      let count = accounts.length;
      let emptyAccounts = [];
      const numEmptyToScan = 10;

      for (; ; count++) {
        const isAccountEmpty = await this.isAccountEmpty(address, ethQuery);
        isAccountEmpty ? emptyAccounts.push(address) : (emptyAccounts = []);
        if (emptyAccounts.length === numEmptyToScan) {
          break;
        }

        ({ addedAccountAddress: address } =
          await this.keyringController.addNewAccount(count));
      }

      // Leave initial account if everything was empty
      const start = count === numEmptyToScan ? 1 : 0;

      for (let i = start; i < emptyAccounts.length; i++) {
        await this.removeAccount(emptyAccounts[i]);
      }
```